src/Microsoft.DotNet.Build.Tasks/VersionTools/SubmitPullRequest.cs

```diff
@@ -0,0 +1,94 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.Build.Framework;
+using Microsoft.DotNet.VersionTools.Automation;
+using Microsoft.DotNet.VersionTools.Automation.GitHubApi;
+using System.Diagnostics;
+using System.Linq;
+
+namespace Microsoft.DotNet.Build.Tasks.VersionTools
+{
+    public class SubmitPullRequest : BuildTask
+    {
+        [Required]
+        public string GitHubAuthToken { get; set; }
+        [Required]
+        public string GitHubUser { get; set; }
+        [Required]
+        public string GitHubEmail { get; set; }
+
+        public string ProjectRepoOwner { get; set; }
+
+        [Required]
+        public string ProjectRepoName { get; set; }
+        [Required]
+        public string ProjectRepoBranch { get; set; }
+
+        public string CommitMessage { get; set; }
+
+        /// <summary>
+        /// Title of the pull request. Defaults to the commit message.
+        /// </summary>
+        public string Title { get; set; }
+
+        /// <summary>
+        /// Body of the pull request. Optional.
+        /// 
+        /// Only used when submitting a new pull request or if TrackDiscardedCommits is false.
+        /// </summary>
+        public string Body { get; set; }
+
+        /// <summary>
+        /// The git author of the update commit. Defaults to the same as GitHubUser.
+        /// </summary>
+        public string GitHubAuthor { get; set; }
+
+        public ITaskItem[] NotifyGitHubUsers { get; set; }
+
+        public bool AlwaysCreateNewPullRequest { get; set; }
+
+        public bool MaintainersCanModifyPullRequest { get; set; }
+
+        public bool TrackDiscardedCommits { get; set; }
+
+        public override bool Execute()
+        {
+            Trace.Listeners.MsBuildListenedInvoke(Log, TraceListenedExecute);
+            return !Log.HasLoggedErrors;
+        }
+
+        private void TraceListenedExecute()
+        {
+            var auth = new GitHubAuth(GitHubAuthToken, GitHubUser, GitHubEmail);
+
+            using (GitHubClient client = new GitHubClient(auth))
+            {
+                var origin = new GitHubProject(ProjectRepoName, GitHubUser);
+
+                var upstreamBranch = new GitHubBranch(
+                    ProjectRepoBranch,
+                    new GitHubProject(ProjectRepoName, ProjectRepoOwner));
+
+                string body = Body ?? string.Empty;
+
+                if (NotifyGitHubUsers != null)
+                {
+                    body += PullRequestCreator.NotificationString(NotifyGitHubUsers.Select(item => item.ItemSpec));
+                }
+
+                var prCreator = new PullRequestCreator(client.Auth, origin, upstreamBranch, GitHubAuthor)
+                {
+                    ForceCreate = AlwaysCreateNewPullRequest,
```

Comment:
Why are these moving to properties on the class?  It seems like all 3 of these should be parameters to `CreateOrUpdateAsync`.

Reply:
I'm not familiar with any guideline people generally use for this--I figured that these bools describe the behavior of the `PullRequestCreator`, and the parameters describe pull request data. Adding many arguments to a function makes me uneasy, but I can't explain that. I don't see a reason for `UpstreamBranch` and `Origin` to be properties if nothing else is. Do you have a methodology for figuring out where to put things?

Reply:
On a class that only has a single method, it is pretty hard to differentiate between where to use properties and where to use parameters to the one method it has.

My suggestion would be to think about if you wanted to create multiple PRs using a single instance of the class.  If it makes sense that every PR created by this class shares these values, then I would say move it to a property.  But if you can imagine a scenario where on PR would want something one way, and another PR would want the option a different way, I'd put the option as a parameter to the method.

Another way of thinking about it - does this parameter only apply to the functionality of the method?

If you are concerned about the number of parameters growing, we can create a "PullRequestOptions" class that encapsulates all the possible options.

Reply:
Thanks, that's very helpful.

For creating multiple PRs, it's really hard to imagine a reason to do that... going with something crazy like submitting for multiple repos, branches, etc. at once, I would expect `_auth` and `GitAuthorName` to be the same: generally I'd say the same person/bot is responsible for all those PRs. Probably nothing else. Good call on the options class.